### ISSUE TITLE
Blazor WASM: "[MONO] * Assertion at ..., condition `<disabled>' not met" (AGAIN)

### ISSUE THREAD

**Issue by PaulSorauer-IPG** [dotnet/runtime#99030]
### Description

This looks like a gnarly one. 😳

I'm getting an "**Error: [MONO] * Assertion at /__w/1/s/src/mono/mono/sgen/sgen-gc.c:2849, condition `<disabled>' not met**" exception raised in the browser in multiple places in my Blazor WASM app.

Anecdotally, this started around the point that I upgraded the .NET version from 8.0.0 to 8.0.2.

I've found a couple of similar threads, but they've long been resolved:
- https://github.com/dotnet/runtime/issues/42628
- https://github.com/dotnet/linker/issues/1507

I was wondering if it was simply an 8.0.2 issue, but to my horror, if I downgrade my .NET version the error still occurs. "And why do you use the word, 'horror'?", I hear you ask? I'm glad you asked...

Even if I'm running .NET 8.00 or .NET 8.0.1, despite cleaning the repo via a `git clean -fXd` command, rebuilding the solution, clearing the entire browser cache or using different browsers, when I right-click the error in the browser it seems to indicate that it's coming from the 8.0.2 runtime:

![2024-02-28_16-01-07](https://github.com/dotnet/runtime/assets/61489185/7c1d0d6d-9561-4f07-b9eb-57db0e3db94b)

I don't understand how I can no reference at all to .NET 8.0.2 with all the previous build artefacts nuked, but the app seems to still be running 8.02.

So long story short, I can't seem to get rid of the exception now it's started. Doesn't matter if I'm running .NET 8.00 or 8.0.1 or 8.0.2.

Can anyone assist me in troubleshooting this further, please? I've absolutely no idea where to go from here.

Cheers.

### Reproduction Steps

I have no idea how to create something to preproduce this, considering I don't know where the problem lies and my entire solution is 150K lines of code (which I can't release).

### Expected behavior

No Mono runtime exceptions

### Actual behavior

Getting Mono runtime exceptions

### Regression?

As above, I'm getting this across versions of .NET 8.0.0 - 8.0.2.

### Configuration

Using Blazor - tried in both Chrome and Edge.

**Comment by PaulSorauer-IPG** [comment]
Actually we found the cause but I still think this is a strange symptom...

We were overriding an async init but had called the synchronous version of the base init by mistake.

We had a large commit so it was hard to find. How that error caused this is a whole different question.

Leaving this open for visibility in case there's a deeper issue.

**Comment by voroninp** [comment]
I get this error both with .NET 8 and .NET 9 preview 6
when I enable IL stripping: `    <WasmStripILAfterAOT>true</WasmStripILAfterAOT>`

![image](https://github.com/dotnet/runtime/assets/9202028/d2a043ca-cf24-4be6-9710-d6bd228b13fe)

[voroninp.github.io-1720597518539.log](https://github.com/user-attachments/files/16157245/voroninp.github.io-1720597518539.log)

**Comment by joer33304** [comment]
Getting this too, I dont have `<WasmStripILAfterAOT>true</WasmStripILAfterAOT>` set

![image](https://github.com/user-attachments/assets/8a0dc982-866f-4f8d-bbd4-6777a9a93bb8)

**Comment by voroninp** [comment]
Still the same error on latest preview .NET 9.0.0-preview.7.24405.7

**Comment by joer33304** [comment]
Error still present and still no resolution, I am coming to the conclusion that Blazor is useless.

**Comment by sofiageo** [comment]
> Error still present and still no resolution, I am coming to the conclusion that Blazor is useless.

Did you try to delete bin / obj directories before trying? Hot reload for simple changes started working again after updating to .NET 9 RC1. But I haven't tried it yet on a larger project.

**Comment by joer33304** [comment]
In my case restarting the app always clears the error. However refreshing the whole app ion the browser does not clear the error

**Comment by Artilizansa** [comment]
Also got the same. Not exactly at the same .c file but the same exception. On my case, I have two repositories implementing an Interface. The first uses IndexeDB and the second, SQLite. The exception occurs only when the apps runs under SQLite. I can't understand the relationship. I have no idea where to start to diag ...

**Comment by mtsfoni** [comment]
I don't know if other people are as stupid as I am:
But I got an error like this when changing target framework to 9 but leaving the `Microsoft.AspNetCore.Components.WebAssembly` and -`.DevServer` library on version 8. Changing those libraries to 9-rc2 fixed my problems (with aot+stripping as well as without).

Maybe it helps somebody who gets lost to this thread.

**Comment by voroninp** [comment]
@mtsfoni , these are my references:
```
<ItemGroup>
  <PackageReference Include="BlazorBeforeUnload" Version="1.*" />
  <PackageReference Include="Blazored.LocalStorage" Version="4.*" />
  <PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly" Version="9.0.0-rc.2.24474.3" />
  <PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly.DevServer" Version="9.0.0-rc.2.24474.3" PrivateAssets="all" />
  <PackageReference Include="Microsoft.Extensions.Localization" Version="9.0.0-rc.2.24474.3" />
  <PackageReference Include="Microsoft.Web.LibraryManager.Build" Version="2.*" />
  <PackageReference Include="System.Private.Uri" Version="4.3.2" />
  <PackageReference Include="System.Text.Json" Version="9.0.0-rc.2.24473.5" />
</ItemGroup>
```

I have issue with .NET 9 only when AOT is enabled. I had no problems with .NET 8

**Comment by voroninp** [comment]
@vitek-karas , what has changed in .NET that AOT fails with regex having option `NoBacktracking`? As soon as I removed it, my app started to work.

**Comment by vitek-karas** [comment]
@voroninp 

Can you please specify exactly how you turn on AOT?

For console apps we have `PublishAOT=true` which will end up using NativeAOT.
On WASM we don't use NativeAOT but the mono AOT.

My suspicion is that with AOT on in WASM we mess up `IsDynamicCodeSupported` somewhere and the regex engine goes down the path which would require interpreter and full metadata present.
Could you please try with .NET 9 GA? There was a [fix](https://github.com/dotnet/runtime/pull/107874) for something similar late in 9 in regex.

(also - maybe create a separate issue for the regex problem as it seems unrelated to the original issue here)

**Comment by voroninp** [comment]
@vitek-karas , for console app it's indeed `<PublishAot>true</PublishAot>`, for Blazor it's `<RunAOTCompilation>true</RunAOTCompilation>`

I'll create a separate issue.

**Comment by BrzVlad** [comment]
The interpreter issue is caused by a missing wrapper in the aot image. It shouldn't really happen for these types of calls. I would need a repro to be able to tell exactly what is going on. An archive with the published app could definitely be useful if possible.

**Comment by actbit** [comment]
I had the same problem when I updated from .net 8 to .net 9, but the following steps solved it for me.

I don't know why it solved it.

1. Change TargetFramework back to .net 8 from Visual Studio properties GUI and build. (You will get an error here, ignore it.)

2. Change TargetFramework to net9.0 in .csproj

3. Build and the problem will be solved.

I hope this helps someone.

**Comment by ruxo** [comment]
I encountered the same issue with my hosted WASM solution, which consists of two projects: Server and Client (WASM).

After upgrading to .NET 9, I observed the same error message.

Initially, I tried deleting the `bin` and `obj` directories in both the Server and Client projects, but that didn’t resolve the problem.

Interestingly, I found a workaround that resolved the issue for me:

1. I created a new hosted WASM solution in .NET 9 and named it `Test9`.
2. I modified the Server project file in `Test9` to reference my original Client project (the one causing the issue) instead of the `Test9` Client project.

After making this change, I cleaned all projects and rebuilt the solution.

To my surprise, the Client started working fine when hosted under this new solution. Afterward, I switched back to my original solution, cleaned and rebuilt it, and it worked smoothly. Even after subsequent clean-ups and rebuilds, the problem did not reoccur.

Based on this experience, I suspect the issue is related to some form of caching, potentially in the workload tooling itself, rather than being linked to the compiled objects.

**Comment by Hillpos** [comment]
When upgrading the version from .net8 to .net9 the error still persists. Has anyone solved the error?

**Comment by nrtncr** [comment]
When upgrading the version from .net8 to .net9 the error still persists. Has anyone solved the error?

**Comment by sofiageo** [comment]
I just tried the 9.0.2XX dotnet sdk daily build and it seems hot reload finally works properly. I hope it will stay like this from now on 👍

**Comment by sascha-wegener** [comment]
I had the same issue with "[MONO] * Assertion at ..., condition `<disabled>' not met" when using `[MethodImpl(MethodImplOptions.Synchronized)]`. Removing it fixed the issue for me.

**Comment by SonOfKorriban** [comment]
> I don't know if other people are as stupid as I am: But I got an error like this when changing target framework to 9 but leaving the `Microsoft.AspNetCore.Components.WebAssembly` and -`.DevServer` library on version 8. Changing those libraries to 9-rc2 fixed my problems (with aot+stripping as well as without).
> 
> Maybe it helps somebody who gets lost to this thread.


After setting the target framework to net9.0, i instantly got that error.
Changing Microsoft.AspNetCore.Components.WebAssembly.DevServer to Version="9.0.6" fixed it for me

`<PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly.DevServer" Version="9.0.6" PrivateAssets="all" />`

**Comment by pavelsavara** [comment]
There are many different problems mixed in the same GH issue here. I'm going to close this one. 
Please create new issues with detailed steps to reproduce, if your problem still persists. 

General hint: delete `bin` and `obj` folders after any .csproj changes, sdk upgrade or workload upgrade.
